text/1004-transactions-api.rst

```diff
@@ -391,6 +391,53 @@ The ``RetryOptions`` signature:
         ): RetryOptions;
     }
 
+Subtransactions API
+'''''''''''''''''''
+
+Subtransaction API is only accessible on the transaction object:
+
+.. code-block:: typescript
+
+    class Transaction {
+        async subtransaction<T>(action: TransactionBlock): Promise<T>;
+    }
+
+Example using ``subtransaction``:
+
+.. code-block:: typescript
+
+    await pool.retry(tx => {
+        try {
+            tx.subtransaction(_ => {
+                tx.query(`
+                    INSERT Note { name := <str>$name, text := <str>$text }
+                `, {name, text})
+            })
+        } catch(e) {
+            if(e instanceof ConstraintViolationError) {
+                tx.subtransaction(tx => {
+                    tx.query(`
+                        UPDATE Note
+                        FILTER .name = <str>$name
+                        SET { text := <str>$text }
+                    `, {name, text})
+                })
+            }
+            raise e
+        }
+    })
+
+Note:
+
+1. Transaction object is not "borrowed" for the duration of subtransaction,
```

Comment:
What does this mean? Does it mean I can do `tx.subtransaction(tx => {` and while the closure is executing use `tx` in some other place?

Reply:
Well, it doesn't make sense to use `tx` in "other place". But you use the same `tx` object for all queries and nested subtransactions. The note here is to show difference between subtransactions and `tryTransaction` / `retryingTransaction`.

Reply:
I'm still uncertain whether we want to pass the same object into subtransaction function in the JavaScript (or have zero-argument function there and rely on closure to propagate `tx` object).

Reply:
> Well, it doesn't make sense to use tx in "other place"

Sure. But it doesn't mean one can't do that by mistake.

> I'm still uncertain whether we want to pass the same object into subtransaction function in the JavaScript (or have zero-argument function there and rely on closure to propagate tx object).

But why don't we create a separate "subtransaction object"? E.g.

```
for tx in con.retry():
   with tx:
      with tx.substransaction() as sub:
         sub.query()
```

?

Reply:
I don't have a good answer to that. I think there is less chance of loosing `tx` to another coroutine, just because it is not common to put transaction object to some global state. Also `tx` expires when transaction is closed. But yes, when the business logic is so large that it requires subtransactions it may have different constraints. Another argument for `transaction.query` is `pool.acquire`, and this argument doesn't apply to subtransactions too.

So this is mostly a matter if we want `subtransaction` to behave similarly.

Reply:
Well I almost forget our discussion about `subtx.rollback()`. So having that we probably need a real subtransaction object.

Reply:
> So having that we probably need a real subtransaction object.

Yeah. @elprans Please read this discussion as it's API-related and confirm.

Reply:
How would `subtransaction()` interact with `DECLARE SAVEPOINT/ROLLBACK TO SAVEPOINT`?  Will we ban them like we did with transaction statements?

As for subtransaction objects and enclosing transaction borrowing, I don't see why we shouldn't follow the same design as for transactions.

Reply:
> How would `subtransaction()` interact with `DECLARE SAVEPOINT/ROLLBACK TO SAVEPOINT`? Will we ban them like we did with transaction statements?

Yes. They are classified as transaction statements now (although banning them is still on to do list)

> As for subtransaction objects and enclosing transaction borrowing, I don't see why we shouldn't follow the same design as for transactions.

Okay.